lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -4563,6 +4563,10 @@ static unsigned AllocateSemanticIndex(
       Type *EltTy = Ty->getStructElementType(i);
       argIdx = AllocateSemanticIndex(EltTy, semIndex, argIdx, endArgIdx,
                                      FlatAnnotationList);
+      // Unwrap array types when checking whether this is a leaf node,
+      // otherwise, array of struct will be misinterpreted as a leaf node.
+      while (EltTy->isArrayTy())
```

Comment:
Similar to this circumstance, I notice that we only ever go one level deep if a struct is detected, and recurse. I would expect something like 
`
 while (EltTy->isStructTy())
        EltTy = EltTy->getStructElementType();
`
but instead we use reccursion. Is it because every struct that is encountered requires an increment in argIdx? If not, why is drilling into an Array treated differently than drilling into a struct here?

Reply:
The logic is a bit complicated, but I think it is sound.

The argument index keeps track of which flattened argument we are currently assigning indices to.  The only time we are changing arguments is when we advance past a struct field of basic type, since each of these are what become separate arguments after flattening.  You don't want to increment on a field of struct type because that will double-increment: an increment from the inner struct field, then an increment from the outer struct field, but the outer field doesn't have an argument of its own, unless the field is a "leaf node", hence the bug.

When we iterate into an array, if it contains a struct, we might have to add indices to multiple final flattened arguments, but since semantic index assignment and increment is depth-first, we have to visit those fields/arguments for each element of the array, adding a few semantic indices at a time across any fields in that element of the array.  This is why the argument index is incremented only when processing a struct inside the array, and returned, but then the argument index starts at the beginning again for each iteration of array elements, finally returning the incremented argument index from the last iteration.  If the array did not contain a struct, no argument increment will occur when handling the array, so if it was a field of a struct, any needed increment will be performed when advancing past the field.

In theory, you could assign blocks of indices without iterating array elements, but it would be even more complicated to compute (requiring index expansion size information from all other struct fields at each level when computing indexing for one field), and you still need to add each index individually, so you can't save much of the work required.

I think the best future improvement might be to abstract the semantic/index assignment logic so that it may use the same code here as we would want to use in Sema to validate the signature elements.